File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/forth/forth.js
//
// This is only a SKELETON file for the 'Forth' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Forth {
  constructor() {
    this._stack = [];
    this.customWords = {};
  }

  evaluate(input) {
    const tokens = input.toUpperCase().split(/\s+/).filter(t => t !== '');
    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      if (token === ':') {
        i++;
        if (i >= tokens.length) {
          throw new Error('Invalid word definition');
        }
        const wordName = tokens[i];
        if (/^\d+$/.test(wordName)) {
          throw new Error('Invalid word name');
        }
        i++;
        const definition = [];
        while (i < tokens.length && tokens[i] !== ';') {
          definition.push(tokens[i]);
          i++;
        }
        if (i >= tokens.length) {
          throw new Error('Unclosed word definition');
        }
        i++;
        this.customWords[wordName.toUpperCase()] = definition;
      } else {
        this.processToken(token);
        i++;
      }
    }
  }

  processToken(token) {
    if (/^-?\d+$/.test(token)) {
      this._stack.push(parseInt(token, 10));
    } else if (this.customWords[token]) {
      this.evaluate(this.customWords[token].join(' '));
    } else {
      this.executeBuiltIn(token);
    }
  }

  executeBuiltIn(token) {
    switch (token) {
      case '+':
        this.ensureStackSize(2);
        this._stack.push(this._stack.pop() + this._stack.pop());
        break;
      case '-':
        this.ensureStackSize(2);
        const sub = this._stack.pop();
        this._stack.push(this._stack.pop() - sub);
        break;
      case '*':
        this.ensureStackSize(2);
        this._stack.push(this._stack.pop() * this._stack.pop());
        break;
      case '/':
        this.ensureStackSize(2);
        const divisor = this._stack.pop();
        if (divisor === 0) {
          throw new Error('Division by zero');
        }
        this._stack.push(Math.trunc(this._stack.pop() / divisor));
        break;
      case 'DUP':
        this.ensureStackSize(1);
        const dup = this._stack.pop();
        this._stack.push(dup, dup);
        break;
      case 'DROP':
        this.ensureStackSize(1);
        this._stack.pop();
        break;
      case 'SWAP':
        this.ensureStackSize(2);
        const a = this._stack.pop();
        const b = this._stack.pop();
        this._stack.push(a, b);
        break;
      case 'OVER':
        this.ensureStackSize(2);
        this._stack.push(this._stack[this._stack.length - 2]);
        break;
      default:
        throw new Error(`Unknown command: ${token}`);
    }
  }

  ensureStackSize(n) {
    if (this._stack.length < n) {
      throw new Error('Stack underflow');
    }
  }

  get stack() {
    return [...this._stack];
  }
}

File: src/2025-02-23-deepclaude-try/go/exercises/practice/forth/forth.go
package forth

import (
	"errors"
	"strconv"
	"strings"
)

// Dictionary stores word definitions
type Dictionary map[string][]string

// Stack operations
func (d Dictionary) evalToken(stack []int, token string) ([]int, error) {
	token = strings.ToUpper(token)
	
	// Try to parse as number
	if num, err := strconv.Atoi(token); err == nil {
		return append(stack, num), nil
	}

	// Check stack has enough elements
	switch token {
	case "+", "-", "*", "/", "SWAP", "OVER":
		if len(stack) < 2 {
			return nil, errors.New("stack underflow")
		}
	case "DUP", "DROP":
		if len(stack) < 1 {
			return nil, errors.New("stack underflow")
		}
	}

	// Handle operations
	switch token {
	case "+":
		stack[len(stack)-2] = stack[len(stack)-2] + stack[len(stack)-1]
		return stack[:len(stack)-1], nil
	case "-":
		stack[len(stack)-2] = stack[len(stack)-2] - stack[len(stack)-1]
		return stack[:len(stack)-1], nil
	case "*":
		stack[len(stack)-2] = stack[len(stack)-2] * stack[len(stack)-1]
		return stack[:len(stack)-1], nil
	case "/":
		if stack[len(stack)-1] == 0 {
			return nil, errors.New("division by zero")
		}
		stack[len(stack)-2] = stack[len(stack)-2] / stack[len(stack)-1]
		return stack[:len(stack)-1], nil
	case "DUP":
		return append(stack, stack[len(stack)-1]), nil
	case "DROP":
		return stack[:len(stack)-1], nil
	case "SWAP":
		stack[len(stack)-2], stack[len(stack)-1] = stack[len(stack)-1], stack[len(stack)-2]
		return stack, nil
	case "OVER":
		return append(stack, stack[len(stack)-2]), nil
	default:
		if def, exists := d[token]; exists {
			// Add tokens to be processed to avoid recursion
			tokens := make([]string, len(def))
			copy(tokens, def)
			return stack, nil
		}
		return nil, errors.New("undefined operation")
	}
}

func Forth(input []string) ([]int, error) {
	dict := make(Dictionary)
	stack := make([]int, 0)
	
	for _, line := range input {
		tokens := strings.Fields(strings.ToUpper(line))
		
		// Handle word definition
		if len(tokens) > 0 && tokens[0] == ":" {
			if len(tokens) < 4 || tokens[len(tokens)-1] != ";" {
				return nil, errors.New("invalid definition")
			}
			name := tokens[1]
			
			// Check if name is a number
			if _, err := strconv.Atoi(name); err == nil {
				return nil, errors.New("illegal operation")
			}
			
			// Expand definition by replacing defined words
			expanded := make([]string, 0)
			for _, t := range tokens[2:len(tokens)-1] {
				if def, exists := dict[t]; exists {
					expanded = append(expanded, def...)
				} else {
					expanded = append(expanded, t)
				}
			}
			
			// Store expanded definition without ":" and ";"
			dict[name] = expanded
			continue
		}

		// Process all tokens iteratively
		tokensToProcess := make([]string, len(tokens))
		copy(tokensToProcess, tokens)
		
		for len(tokensToProcess) > 0 {
			token := tokensToProcess[0]
			tokensToProcess = tokensToProcess[1:]
			
			// If token is a defined word, prepend its definition to tokens
			if def, exists := dict[token]; exists {
				newTokens := make([]string, len(def)+len(tokensToProcess))
				copy(newTokens, def)
				copy(newTokens[len(def):], tokensToProcess)
				tokensToProcess = newTokens
				continue
			}
			
			// Evaluate the token
			var err error
			stack, err = dict.evalToken(stack, token)
			if err != nil {
				return nil, err
			}
		}
	}

	return stack, nil
}
